src/System.Private.CoreLib/shared/System/TimeZoneInfo.Unix.cs

```diff
@@ -413,24 +415,34 @@ private static IEnumerable<string> EnumerateFilesRecursively(string path)
         {
             List<string> toExplore = null; // List used as a stack
 
+            int bufferSize = Interop.Sys.GetReadDirRBufferSize();
             string currentPath = path;
             for(;;)
             {
-                using (Microsoft.Win32.SafeHandles.SafeDirectoryHandle dirHandle = Interop.Sys.OpenDir(currentPath))
+                IntPtr dirHandle = Interop.Sys.OpenDir(currentPath);
+                if (dirHandle == IntPtr.Zero)
                 {
-                    if (dirHandle.IsInvalid)
-                    {
-                        throw Interop.GetExceptionForIoErrno(Interop.Sys.GetLastErrorInfo(), currentPath, isDirectory: true);
-                    }
+                    throw Interop.GetExceptionForIoErrno(Interop.Sys.GetLastErrorInfo(), currentPath, isDirectory: true);
+                }
 
+                byte[] dirBuffer = null;
+                GCHandle dirBufferHandle = default;
+                try
+                {
+                    dirBuffer = ArrayPool<byte>.Shared.Rent(bufferSize);
+                    dirBufferHandle = GCHandle.Alloc(dirBuffer, GCHandleType.Pinned);
                     // Read each entry from the enumerator
                     Interop.Sys.DirectoryEntry dirent;
-                    while (Interop.Sys.ReadDir(dirHandle, out dirent) == 0)
+                    while (Interop.Sys.ReadDirR(dirHandle, dirBufferHandle.AddrOfPinnedObject(), bufferSize, out dirent) == 0)
                     {
-                        if (dirent.InodeName == "." || dirent.InodeName == "..")
+                        Span<char> nameBuffer = stackalloc char[Interop.Sys.DirectoryEntry.NameBufferSize];
+                        ReadOnlySpan<char> direntName = dirent.GetName(nameBuffer);
+
+                        if (((uint)direntName.Length == 1u && direntName[0] == '.') ||
```

Comment:
Is this casting to `uint` really needed? As far as I can tell, the code is identical to the one without casting with current RyuJIT.

Reply:
paranoia 🙂 just checked output - casts are not needed here